### PR TITLE
Add custom link and image rendering functions

### PR DESCRIPTION
**Pull request by m-murasovs**
Needed for: https://app.zenhub.com/workspaces/platform-team-5f6454160d9f82000fa6733f/issues/apify/apify-web/861

### CODE REVIEW

File: src/markdown_renderers.js [hunk 1]
```diff
@@ -49,3 +52,69 @@ export const customHeadingRenderer = (text, level, raw) => {
...
+    if (href.startsWith('#')) {
+        href = href.toLowerCase();
+    }
+
+    // Replace relative URLs in README with absolute ones pointing to the actor's repo
+    if (!href.startsWith('http') && !href.startsWith('ftp') && !href.startsWith('#')) {
+        href = `${urlPrefix}/${href}`;
+    }
+
+    return `<a href=${href}>${text}</a>`;
```
Comment (drobnikj) [inline src/markdown_renderers.js:98]: You should to add rel="nofollow noreferrer" att into links go out of apify.

**Review by drobnikj — CHANGES_REQUESTED** [review]

File: src/markdown_renderers.js [hunk 2]
```diff
@@ -49,3 +52,69 @@ export const customHeadingRenderer = (text, level, raw) => {
...
+ * Mainly for use in actor READMES
+ * Parses the actor's repo URL to extract the repo name and owner name.
+ * @param {string} href
+ * @param {string} text
+ * @param {string} repoUrl
+ * @param {string} branchName
+ * @return {string}
+*/
+export const customLinkRenderer = (href, text, repoUrl, branchName) => {
+    const urlPrefix = generateGitRepoUrlPrefix(repoUrl, branchName);
```
[the remark this answers is not in the document]
Comment (m-murasovs) [reply src/markdown_renderers.js:86]: I wanted to do this initially but the github URL uses either `/tree` if you're linking a folder or `/blob` if you're linking a file, which would need some additional parsing. Do you think it's worth it?

File: src/markdown_renderers.js [hunk 3]
```diff
@@ -1,3 +1,6 @@
+function interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } // eslint-disable-line
```
[the remark this answers is not in the document]
Comment (m-murasovs) [reply src/markdown_renderers.js:1]: Yeah, I wasn't sure about this one, I was looking at the built `git` file for inspiration and it had this. I'll revise

File: src/markdown_renderers.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-murasovs) [reply src/markdown_renderers.js:86]: OK, I did it :D

File: src/markdown_renderers.js [hunk 4]
```diff
@@ -49,3 +51,97 @@ export const customHeadingRenderer = (text, level, raw) => {
...
+ * @param {string} repoUrl
+ * @param {string} branchName
+ * @return {string}
+*/
+export const customLinkRenderer = (href, text, repoUrl, branchName) => {
+    // Ensure that anchors have lowercase href
+    if (href.startsWith('#')) {
+        href = href.toLowerCase();
+    }
+    // Only target relative URLs, which are used to refer to the git repo, and not anchors or absolute URLs
```
Comment (nguyeda1) [inline src/markdown_renderers.js:121]: Is this enough to check if href is relative URL?

Comment (m-murasovs) [reply src/markdown_renderers.js:121]: Yeah, you're right! Fix in

File: src/markdown_renderers.js [hunk 5]
```diff
@@ -49,3 +51,98 @@ export const customHeadingRenderer = (text, level, raw) => {
...
+    const relativeUrlRegex = new RegExp('^(?:[a-z]+:)?//', 'i');
+    if (!relativeUrlRegex.test(href)) {
```
Comment (nguyeda1) [inline src/markdown_renderers.js:123]: I guess `relativeUrlRegex ` is regex for absolute URLs right? therefore it should be named `absoluteUrlRegex`.. also it looks like it fails if the URL is eg www.google.com see https://regex101.com/r/alX8HI/1

File: src/markdown_renderers.js [hunk 6]
```diff
@@ -49,3 +51,98 @@ export const customHeadingRenderer = (text, level, raw) => {
...
+ * Mainly for use in actor READMES
+ * Parses the actor's repo URL to extract the repo name and owner name.
+ * @param {string} href
+ * @param {string} text
+ * @param {string} repoUrl
+ * @param {string} branchName
+ * @return {string}
+*/
+export const customImageRenderer = (href, text, repoUrl, gitBranchName) => {
+    if (!href.startsWith('http') && !href.startsWith('ftp') && !href.startsWith('#')) {
```
Comment (nguyeda1) [inline src/markdown_renderers.js:142]: This should be regex tested as well

**Review by nguyeda1 — CHANGES_REQUESTED** [review]

**Review by nguyeda1 — CHANGES_REQUESTED** [review]
One last thing otherwise looks good :) I would create a separate function `isUrlRelative` somewhere in utilities probably for checking the URL.. it might be useful in other usecases too :)

File: src/markdown_renderers.js [hunk 7]
```diff
@@ -1,3 +1,7 @@
+import gitUrlParse from 'git-url-parse';
+
+const relativeUrlRegex = new RegExp('^(?!www.|(?:http|ftp)s?://|[A-Za-z]:\\|//).*', 'i');
```
Comment (jbartadev) [inline src/markdown_renderers.js:3]: Maybe put this to `src/regexs.js`

File: src/markdown_renderers.js [hunk 8]
```diff
@@ -49,3 +53,98 @@ export const customHeadingRenderer = (text, level, raw) => {
             <h${level} id="${headingId}"><a href="#${headingId}"></a>${headingText}</h${level}>`;
     return headingToReturn;
 };
+
+export const parseRepoName = (repoUrl) => {
+    // Can't use parsedRepoUrl.full_name on it's own as Bitbucket adds irrelevant path suffix to the end of it
+    const parsedRepoUrl = gitUrlParse(repoUrl);
+    const repoNameParts = parsedRepoUrl.full_name.split('/');
+    const repoFullName = `${repoNameParts[0]}/${repoNameParts[1]}`;
+    return repoFullName;
```
Comment (jbartadev) [inline src/markdown_renderers.js:62]: Maybe something like this?
```
const parsedRepoUrl = gitUrlParse(repoUrl);
const repoName = parsedRepoUrl.full_name.split('/').slice(0,2).join('/');
return repoName;
```
The parse`RepoName` func name and the repo`FullName` return value don't match so it makes sense to do it like this. Or just return the result of the last command.
I don't really know how many parts are there in `repoNameParts` but probably still makes sense to use the slice and join...

File: src/markdown_renderers.js [hunk 9]
```diff
@@ -49,3 +53,98 @@ export const customHeadingRenderer = (text, level, raw) => {
...
+    const repoFullName = parseRepoName(repoUrl);
+
+    const hrefParts = href.split('/');
+    const lastHrefPart = hrefParts[hrefParts.length - 1];
+
+    // If the last part of the URL has a dot, it's a file with an extension or .gitignore (blob),
+    // otherwise we assume the link is for a directory (tree)
+    const isTreeOrBlob = lastHrefPart.includes('.') ? 'blob' : 'tree';
+
+    if (repoUrl.includes('github.com')) {
```
Comment (jbartadev) [inline src/markdown_renderers.js:97]: It would be nice to have regexes for this too, probably not needed now

**Review by jbartadev — CHANGES_REQUESTED** [review]
Looks good, just a minor comments

**Review by jbartadev — APPROVED** [review]
Looks good

**Review by nguyeda1 — APPROVED** [review]

**Pull request merged**